### PR TITLE
Add proper error messages for submitJob command

### PR DESCRIPTION
**Pull request by zeineb**

### CODE REVIEW

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java [hunk 1]
```diff
@@ -83,9 +88,60 @@ public void execute(ApplicationContext currentContext) throws CLIException {
...
+            throw new CLIException(REASON_FILE_EMPTY, String.format("'%s' is empty.", pathname));
+        }
+
     }
 
     private Map<String, String> map(String variables) {
         return jobKeyValueTransformer.transformVariablesToMap(variables);
     }
+
+    private Boolean checkFileNotEmpty(String pathname) {
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: can you rename this to isFileNotEmpty , the check doesn't indicate when it returns true vs false

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java [hunk 2]
```diff
@@ -83,9 +88,60 @@ public void execute(ApplicationContext currentContext) throws CLIException {
...
+                return false;
+            } else
+                return true;
+        } catch (IOException e) {
+            e.printStackTrace();
+            return false;
+        }
+    }
+
+    private Boolean checkFileExists(String pathname) {
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: Here similar, can you replace the check with "is"

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java [hunk 3]
```diff
@@ -83,9 +88,60 @@ public void execute(ApplicationContext currentContext) throws CLIException {
...
+        File file = new File(pathname);
+        if (file.exists() && !file.isDirectory()) {
+            return true;
+        } else
+            return false;
+    }
+
+    private Boolean checkFilePathValid(String pathname) {
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: Here similar can you replace the check with "is"

File: rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+		Mockito.when(schedulerRestClientMock.submitJobArchive("sessionid", fileInputStreamMock,
+				jobKeyValueTransormerMock.transformVariablesToMap(""))).thenReturn(null);
+
+		new SubmitJobCommand(params).execute(currentContextMock);
+
+		Mockito.verify(schedulerRestClientMock).submitJobArchive("sessionid", null, null);
+	}
+
+	@Test(expected = CLIException.class)
+	public void testInvalidFilePathProvided() throws Exception {
```
Comment (tobwiens) [inline rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java]: You can improve the naming of the test to something like testInvalidFilePathProvidedThrowsException

File: rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+		assertThat(stack.get(0).getMessage(), is("'" + params[0] + "' is not a valid file."));
+
+		Mockito.verify(schedulerRestClientMock, times(0)).submitJobArchive(anyString(), anyObject(), anyObject());
+
+		throw stack.get(0);
+	}
+
+	@Test(expected = CLIException.class)
+	public void testNonExistingFilePathProvided() throws Exception {
```
Comment (tobwiens) [inline rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java]: You can improve the naming

File: rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+		assertThat(stack.get(0).getMessage(), is("'" + params[0] + "' does not exist."));
+
+		Mockito.verify(schedulerRestClientMock, times(0)).submitJobArchive(anyString(), anyObject(), anyObject());
+
+		throw stack.get(0);
+	}
+
+	@Test(expected = CLIException.class)
+	public void testFileIsEmpty() throws Exception {
```
Comment (tobwiens) [inline rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java]: Here you can improve the naming as well.

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 4 issues

1. CRITICAL [SubmitJobCommand.java#L122](https://github.com/ow2-proactive/scheduling/blob/efff634b866f7e0283d91545755234c01371c291/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L122): Use a logger to log this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1148)
1. MAJOR [SubmitJobCommand.java#L121](https://github.com/ow2-proactive/scheduling/blob/efff634b866f7e0283d91545755234c01371c291/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L121): Either log or rethrow this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. MINOR [SubmitJobCommand.java#L117](https://github.com/ow2-proactive/scheduling/blob/efff634b866f7e0283d91545755234c01371c291/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L117): Replace this if-then-else statement by a single return statement. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1126)
1. MINOR [SubmitJobCommand.java#L129](https://github.com/ow2-proactive/scheduling/blob/efff634b866f7e0283d91545755234c01371c291/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L129): Replace this if-then-else statement by a single return statement. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1126)

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/CommandSet.java [hunk 7]
```diff
@@ -262,10 +262,10 @@
                                                                           .opt("lj")
                                                                           .longOpt("listjobs")
                                                                           .description("Retrieve a list of jobs managed by the Scheduler.If specified the latest 'x' number of jobs or 'z' maximum number of jobs starting from 'y'th job.") //Example: to return the five last jobs, >./proactive-client -lj limit=5. To return the list of z jobs from index y, >./proactive-client -lj from=y limit=z ")
```
Comment (mbenguig) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/CommandSet.java]: Use latest=x to filter the x latest jobs, or from=y limit=z to retrieve the z latest jobs from the yth.

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java (hunk 1 above)
Comment (zeineb) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: I renamed the method name, thanks !

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java [hunk 8]
```diff
@@ -49,22 +53,23 @@
 
     private final JobKeyValueTransformer jobKeyValueTransformer;
 
-    public SubmitJobCommand(String... params) {
-
+    public SubmitJobCommand(String... params) throws CLIException {
+        if (params == null || params.length == 0) {
```
Comment (mbenguig) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: Cant we just test "params.length == 0" ? (this is a question)

Comment (zeineb) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: We need also to check if the object "params" is not null. If we test only on the length, we will get NullPointerException if "params" is null.

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java [hunk 9]
```diff
@@ -83,9 +88,60 @@ public void execute(ApplicationContext currentContext) throws CLIException {
...
     private Map<String, String> map(String variables) {
         return jobKeyValueTransformer.transformVariablesToMap(variables);
     }
+
+    private Boolean checkFileNotEmpty(String pathname) {
+        try {
+            BufferedReader reader = new BufferedReader(new FileReader(pathname));
+            if (checkFileExists(pathname) && reader.readLine() == null) {
+                return false;
+            } else
```
Comment (mbenguig) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: we enter in the else condition when 
!(checkFileExists(pathname) && reader.readLine() == null) <=>
!checkFileExists(pathname)  ||  !(reader.readLine() == null)

can return true if   !checkFileExists(pathname)   (<>checkFileNotEmpty)

Comment (mbenguig) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: Moreover, i would dissociate the file existence test from the  checkFileNotEmpty function.. or i would have renamed it

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java (hunk 2 above)
Comment (mbenguig) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: Can be confusing.. or rename it to   isAnExistingFile

File: rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/CommandSetTest.java [hunk 10]
```diff
@@ -79,7 +79,11 @@ public void testThatArgNamesMatchNumberOfArgs() throws ParseException, IllegalAc
 
             if (!argNames.trim().isEmpty()) {
                 nbArgsBasedOnName = argNames.split(" ").length;
-                nbArgsBasedOnName = nbArgsBasedOnName - optionals;
+                if (entry.hasArgs()) {
```
Comment (mbenguig) [inline rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/CommandSetTest.java]: (Question) Do we consider in this if-condition cases with non optional+ optional args?

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/CommandSet.java (hunk 7 above)
Comment (zeineb) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/CommandSet.java]: Much better !

File: rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+		new SubmitJobCommand(params).execute(currentContextMock);
+
+		Mockito.verify(schedulerRestClientMock).submitJobArchive("sessionid", null, null);
+	}
+
+	@Test(expected = CLIException.class)
+	public void testInvalidFilePathProvided() throws Exception {
+
+		params = new String[1];
+		params[0] = "/test/a.xm";
```
Comment (mbenguig) [inline rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java]: You could be more explicit with the filename:  /non/valid/file.xml

File: rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+		Mockito.verify(schedulerRestClientMock, times(0)).submitJobArchive(anyString(), anyObject(), anyObject());
+
+		throw stack.get(0);
+	}
+
+	@Test(expected = CLIException.class)
+	public void testNonExistingFilePathProvided() throws Exception {
+
+		params = new String[1];
+		params[0] = "/Users/c.xml";
```
Comment (mbenguig) [inline rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java]: same remark

File: rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+		Mockito.verify(schedulerRestClientMock, times(0)).submitJobArchive(anyString(), anyObject(), anyObject());
+
+		throw stack.get(0);
+	}
+
+	@Test(expected = CLIException.class)
+	public void testFileIsEmpty() throws Exception {
+
+		params = new String[1];
+		params[0] = "/Users/zeinebzhioua/Downloads/a.xml";
```
Comment (mbenguig) [inline rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java]: (to check)

Comment (zeineb) [inline rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java]: Still investigating ...

File: rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java (hunk 12 above)
Comment (zeineb) [inline rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java]: This one is ok I think. I can amen it to "/aaa/bbb/c.xml" for example.

File: rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java (hunk 11 above)
Comment (zeineb) [inline rest/rest-cli/src/test/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommandTest.java]: Done.

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java (hunk 2 above)
Comment (zeineb) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: I changed it to "... isFileExisting(...)". Is it still confusing?

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java (hunk 9 above)
Comment (zeineb) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: Addressed. Thanks a lot !

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java [hunk 14]
```diff
@@ -49,22 +53,23 @@
...
 
     @Override
     public void execute(ApplicationContext currentContext) throws CLIException {
-        File jobFile = new File(pathname);
-        if (!jobFile.exists()) {
-            throw new CLIException(REASON_INVALID_ARGUMENTS, String.format("'%s' does not exist.", pathname));
-        }
         try {
+            validateFilePath(currentContext);
+            File jobFile = new File(pathname);
```
Comment (zeineb) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java]: I tried to do so, but the specific CLI exceptions are no longer captured, and only the generic one is raised.

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 2 issues

1. MAJOR [SubmitJobCommand.java#L125](https://github.com/ow2-proactive/scheduling/blob/8fad255588dd243a5c615e8a48a243e67988fef1/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L125): Either log or rethrow this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. MINOR [SubmitJobCommand.java#L133](https://github.com/ow2-proactive/scheduling/blob/8fad255588dd243a5c615e8a48a243e67988fef1/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L133): Replace this if-then-else statement by a single return statement. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1126)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 2 issues

1. MAJOR [SubmitJobCommand.java#L125](https://github.com/ow2-proactive/scheduling/blob/d35ad61e0d46f01e97273faece309e81475c7d05/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L125): Either log or rethrow this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. MINOR [SubmitJobCommand.java#L133](https://github.com/ow2-proactive/scheduling/blob/d35ad61e0d46f01e97273faece309e81475c7d05/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L133): Replace this if-then-else statement by a single return statement. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1126)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 2 issues

1. MAJOR [SubmitJobCommand.java#L125](https://github.com/ow2-proactive/scheduling/blob/f93126d010d0626a84c866f0768f29c64d3d3027/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L125): Either log or rethrow this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. MINOR [SubmitJobCommand.java#L133](https://github.com/ow2-proactive/scheduling/blob/f93126d010d0626a84c866f0768f29c64d3d3027/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L133): Replace this if-then-else statement by a single return statement. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1126)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 2 issues

1. CRITICAL [SubmitJobCommand.java#L126](https://github.com/ow2-proactive/scheduling/blob/7a95ac9c5f74394f5bb951636a0c8b057fdc226d/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L126): Use a logger to log this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1148)
1. MAJOR [SubmitJobCommand.java#L125](https://github.com/ow2-proactive/scheduling/blob/7a95ac9c5f74394f5bb951636a0c8b057fdc226d/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L125): Either log or rethrow this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 1 issue

1. MINOR [SubmitJobCommand.java#L128](https://github.com/ow2-proactive/scheduling/blob/ababae6ed8426e868600a5475f73312e89bc9306/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/SubmitJobCommand.java#L128): Make the "logger" logger private static final. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1312)

**Pull request merged**